### DISCUSSION TITLE
v0.1.1

### DISCUSSION THREAD

**Discussion by 77axel** [77AXEL/PyCNN#7] [General]
## 📦 Release: `v0.1.1 – Training Visualization Added`

This update enhances **CNN-FS (Convolutional Neural Network From Scratch)** by adding **real-time training visualization** for **accuracy and loss per epoch** using Matplotlib — making it easier to monitor model performance while training.

---

### 🚀 Key Features

* ✅ Fully functional CNN implementation from scratch
* 🧠 Manual convolution, max pooling, and ReLU activations
* 🔁 Forward and backward propagation with mini-batch gradient descent
* 🏷 Multi-class classification via softmax and cross-entropy loss
* 💾 Model save/load using `pickle`
* 🖼 RGB image preprocessing with customizable filters
* 🔍 Predict function to classify new unseen images
* 📊 **New:** Real-time training visualization (accuracy & loss per epoch) in `train_model(visualize=True)`

---

#### 📦 Install directly from GitHub:

```bash
pip install git+https://github.com/77AXEL/CNN-FS.git@v0.1.1
```

---

#### 🏁 Train the model with visualization:

```python
from cnnfs.model import CNN

model = CNN()
model.init(
    image_size=32,
    batch_size=32,
    h1=128,
    h2=64,
    learning_rate=0.01,
    epochs=10,
    dataset_path="data",
    max_image=200
)
model.load_dataset()
model.train_model(visualize=True)  # <-- visualization enabled (Without this setting, the visualization won’t appear)
model.save_model()
```

#### 🔍 Predict an image:

```python
model.load_model("model.bin")
result = model.predict("path/to/image.png")
print("Prediction:", result)
```

---

### 🗂 Included Filters

CNN-FS uses a set of basic filters for feature extraction:

* Sharpen
* Vertical edges
* Laplacian (high-pass)

---

#### **v0.1.1**

* **New:** Real-time training visualization (accuracy & loss per epoch) with Matplotlib.
* Dark theme visualization with clear, color-coded curves.
* Optional: Enable by calling `train_model(visualize=True)`.

#### **v0.1.0**

* Initial version with full training and prediction pipeline
* Manual convolution + pooling + ReLU
* Batch softmax + cross-entropy loss
* Two hidden layers and output layer
* Save/load model capability
* Example dataset folder included
* Beginner-friendly structure for learning how CNNs work

---

### 💬 Feedback & Contributions

We welcome issues, suggestions, and contributions!
Join the conversation in the \[[[Discussions tab](https://github.com/77AXEL/CNN-FS/discussions)](https://github.com/77AXEL/CNN-FS/discussions)]\([https://github.com/77AXEL/CNN-FS/discussions](https://github.com/77AXEL/CNN-FS/discussions)) or check the \[[[CONTRIBUTING.md](https://github.com/77AXEL/CNN-FS/?tab=coc-ov-file)](https://github.com/77AXEL/CNN-FS/?tab=coc-ov-file)]\(./CONTRIBUTING.md) guide.

---

### 🛡 Security

Found a security issue? Please report it privately to:
📧 \[[[a.x.e.l777444000@gmail.com](mailto:a.x.e.l777444000@gmail.com)](mailto:a.x.e.l777444000@gmail.com)]

---

### 📜 License

Released under the \[[[MIT License](https://chatgpt.com/c/LICENSE)](./LICENSE)]\(./LICENSE)

<hr /><em>This discussion was created from the release <a href='https://github.com/77AXEL/CNN-FS/releases/tag/v0.1.1'>v0.1.1</a>.</em>
